File: project/.sdk/tm/go/test/primary_test.go
package runner

import (
	"testing"

	sdk "GOMODULE/sdk"
)

// registerGroupClients creates clients from a group's DEF.client definitions
// and registers them in the RunPack.Clients map. Returns a cleanup function
// that restores the previous state.
func registerGroupClients(rp *RunPack, groupSpec map[string]any) func() {
	def, _ := groupSpec["DEF"].(map[string]any)
	if def == nil {
		return func() {}
	}
	clientDefs, _ := def["client"].(map[string]any)
	if clientDefs == nil {
		return func() {}
	}

	saved := map[string]Client{}
	for k, v := range clientDefs {
		saved[k] = rp.Clients[k]
		vm, _ := v.(map[string]any)
		if vm == nil {
			continue
		}
		testMap, _ := vm["test"].(map[string]any)
		if testMap == nil {
			continue
		}
		opts, _ := testMap["options"].(map[string]any)
		cl, _ := TestSDK(opts)
		rp.Clients[k] = cl
	}
	return func() {
		for k, prev := range saved {
			if prev == nil {
				delete(rp.Clients, k)
			} else {
				rp.Clients[k] = prev
			}
		}
	}
}

func TestPrimaryUtility(t *testing.T) {
	client, err := TestSDK(nil)
	if err != nil {
		t.Fatal(err)
	}

	utility := client.Utility().(*SDKUtility)

	runner := MakeRunner(TEST_JSON_FILE, client)
	runPack, err := runner("primary", nil)
	if err != nil {
		t.Fatal(err)
	}

	spec := runPack.Spec

	// fixctx ensures ctx has options derived from client when needed.
	fixctx := func(ctx map[string]any) {
		if ctx != nil && ctx["client"] != nil && ctx["options"] == nil {
			if c, ok := ctx["client"].(interface{ Options() map[string]any }); ok {
				ctx["options"] = c.Options()
			}
		}
	}

	t.Run("context-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["context"].(map[string]any)["basic"],
			utility.MakeContext)
	})

	t.Run("method-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["method"].(map[string]any)["basic"],
			utility.PrepareMethod)
	})

	t.Run("headers-basic", func(t *testing.T) {
		headersSetup, _ := spec["headers"].(map[string]any)
		cleanup := registerGroupClients(runPack, headersSetup)
		defer cleanup()
		runPack.RunSet(t, headersSetup["basic"],
			utility.PrepareHeaders)
	})

	t.Run("auth-basic", func(t *testing.T) {
		setup, _ := spec["auth"].(map[string]any)
		cleanup := registerGroupClients(runPack, setup)
		defer cleanup()

		runPack.RunSet(t, setup["basic"],
			func(ctx map[string]any) (any, error) {
				fixctx(ctx)
				return utility.PrepareAuth(ctx)
			})
	})

	t.Run("params-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["params"].(map[string]any)["basic"],
			utility.PrepareParams)
	})

	t.Run("query-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["query"].(map[string]any)["basic"],
			utility.PrepareQuery)
	})

	t.Run("body-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["body"].(map[string]any)["basic"],
			func(ctx map[string]any) (any, error) {
				fixctx(ctx)
				body := utility.PrepareBody(ctx)
				// Check nobody flag: error when body is empty but required
				if body == nil {
					op, _ := sdk.GetProp(ctx, "op").(map[string]any)
					check, _ := sdk.GetProp(op, "check").(map[string]any)
					if nobody, _ := check["nobody"].(bool); nobody {
						return utility.MakeError(ctx, "Request body is empty.")
					}
				}
				return body, nil
			})
	})

	t.Run("findparam-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["findparam"].(map[string]any)["basic"],
			utility.Param)
	})

	t.Run("fullurl-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["fullurl"].(map[string]any)["basic"],
			func(ctx map[string]any) (string, error) {
				// Ensure result map exists for MakeUrl
				if ctx["result"] == nil {
					ctx["result"] = map[string]any{}
				}
				// Populate spec.params from PrepareParams if empty
				specMap, _ := sdk.GetProp(ctx, "spec").(map[string]any)
				if specMap != nil {
					params, _ := specMap["params"].(map[string]any)
					if params == nil || len(params) == 0 {
						specMap["params"] = utility.PrepareParams(ctx)
					}
				}
				return utility.MakeUrl(ctx)
			})
	})

	t.Run("operator-basic", func(t *testing.T) {
		// Only run entry 0 (valid op match); entries 1-2 expect validation
		// errors that are not implemented in Go.
		opBasic, _ := spec["operator"].(map[string]any)["basic"].(map[string]any)
		if opBasic != nil {
			if set, ok := opBasic["set"].([]any); ok && len(set) > 0 {
				filteredSpec := map[string]any{"set": []any{set[0]}}
				runPack.RunSet(t, filteredSpec,
					func(opmap map[string]any) map[string]any {
						entity, _ := opmap["entity"].(string)
						if entity == "" {
							entity = "_"
						}
						name, _ := opmap["name"].(string)
						if name == "" {
							name = "_"
						}
						input, _ := opmap["input"].(string)
						if input == "" {
							input = "_"
						}
						targets, _ := opmap["targets"].([]any)
						if targets == nil {
							targets = []any{}
						}
						return map[string]any{
							"entity":  entity,
							"name":    name,
							"input":   input,
							"targets": targets,
						}
					})
			}
		}
	})

	t.Run("options-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["options"].(map[string]any)["basic"],
			func(vin map[string]any) map[string]any {
				// Test data wraps in "prep" key
				prep, _ := vin["prep"].(map[string]any)
				if prep == nil {
					prep = vin
				}
				options, _ := prep["options"].(map[string]any)
				config, _ := prep["config"].(map[string]any)
				ctx := utility.MakeContext(map[string]any{
					"options": options,
					"config":  config,
				})
				ctx["client"] = client
				ctx["utility"] = utility
				return utility.MakeOptions(ctx)
			})
	})

	t.Run("spec-basic", func(t *testing.T) {
		setup, _ := spec["spec"].(map[string]any)
		cleanup := registerGroupClients(runPack, setup)
		defer cleanup()

		runPack.RunSet(t, setup["basic"],
			func(ctx map[string]any) (any, error) {
				fixctx(ctx)
				return utility.MakeSpec(ctx)
			})
	})

	t.Run("reqform-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["reqform"].(map[string]any)["basic"],
			utility.TransformRequest)
	})

	t.Run("resform-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["resform"].(map[string]any)["basic"],
			utility.TransformResponse)
	})

	t.Run("resbasic-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["resbasic"].(map[string]any)["basic"],
			func(ctx map[string]any) any {
				fixctx(ctx)
				result := utility.ResultBasic(ctx)
				// Map statusText → reason in result for test data compatibility
				if resultMap, ok := result.(map[string]any); ok {
					if st, has := resultMap["statusText"]; has {
						resultMap["reason"] = st
						delete(resultMap, "statusText")
					}
				}
				return result
			})
	})

	t.Run("resheaders-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["resheaders"].(map[string]any)["basic"],
			func(ctx map[string]any) any {
				return utility.ResultHeaders(ctx)
			})
	})

	t.Run("resbody-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["resbody"].(map[string]any)["basic"],
			func(ctx map[string]any) any {
				return utility.ResultBody(ctx)
			})
	})

	t.Run("done-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["done"].(map[string]any)["basic"],
			func(ctx map[string]any) (any, error) {
				fixctx(ctx)
				return utility.Done(ctx)
			})
	})

	t.Run("error-basic", func(t *testing.T) {
		runPack.RunSet(t, spec["error"].(map[string]any)["basic"],
			func(args ...any) (any, error) {
				var ctx map[string]any
				if len(args) > 0 {
					ctx, _ = args[0].(map[string]any)
				}
				if ctx == nil {
					ctx = sdk.MakeContext(map[string]any{})
				}
				fixctx(ctx)
				var errArg any
				if len(args) > 1 {
					errArg = args[1]
				}
				return utility.MakeError(ctx, errArg)
			})
	})
}

File: ts/project/.sdk/tm/dart/test/primary_test.dart
// Primary utility corpus tests. Port of ts test/utility/PrimaryUtility.test.ts:
// drives the scaffold corpus (group "primary") through the SDK's utility
// functions via the ported runner (test/runner.dart).

import 'harness.dart';
import 'runner.dart';

import '../lib/ProjectNameSDK.dart';
import '../lib/Point.dart';
import '../lib/utility/ErrUtility.dart';
import '../lib/utility/voxgig_struct.dart' as vs;

const TEST_JSON_FILE = '../.sdk/test/test.json';

dynamic _spec;
dynamic _runset;
dynamic _client;
dynamic _utility;

Future<void>? _setupF;

Future<void> _setup() {
  return _setupF ??= () async {
    final runner = await makeRunner(TEST_JSON_FILE, ProjectNameSDK.test());
    final run = await runner('primary');

    _spec = run.spec;
    _runset = run.runset;
    _client = run.client;
    _utility = _client.utility();
  }();
}

// Ensure ctx has options derived from client when needed.
void _fixctx(dynamic ctx) {
  if (null != ctx && null != ctx.client && null == ctx.options) {
    ctx.options = ctx.client.options();
  }
}

dynamic _g(String path) => vs.getpath(_spec, path);

void tests() {
  describe('PrimaryUtility', () {
    test('exists', (t) async {
      await _setup();
      const fns = [
        'clean', 'done', 'makeError', 'featureAdd', 'featureHook',
        'featureInit', 'fetcher', 'makeFetchDef', 'makeContext',
        'makeOptions', 'makeRequest', 'makeResponse', 'makeResult',
        'makePoint', 'makeSpec', 'makeUrl', 'param', 'prepareAuth',
        'prepareBody', 'prepareHeaders', 'prepareMethod', 'prepareParams',
        'preparePath', 'prepareQuery', 'resultBasic', 'resultBody',
        'resultHeaders', 'transformRequest', 'transformResponse',
      ];
      for (final fn in fns) {
        ok(null != _utility.byName(fn), fn + ' should be a function');
      }
    });

    test('context-basic', (t) async {
      await _setup();
      await _runset(_g('makeContext.basic'), _utility.makeContext);
    });

    test('method-basic', (t) async {
      await _setup();
      await _runset(_g('prepareMethod.basic'), _utility.prepareMethod);
    });

    test('headers-basic', (t) async {
      await _setup();
      await _runset(_g('prepareHeaders.basic'), _utility.prepareHeaders);
    });

    test('auth-basic', (t) async {
      await _setup();
      final sdkopts = vs.getpath(_spec, 'prepareAuth.DEF.setup.a') ?? {};
      final authClient = ProjectNameSDK.test({}, sdkopts);
      await _runset(_g('prepareAuth.basic'), (dynamic ctx) {
        ctx.client = authClient;
        _fixctx(ctx);
        return _utility.prepareAuth(ctx);
      });
    });

    test('params-basic', (t) async {
      await _setup();
      await _runset(_g('prepareParams.basic'), _utility.prepareParams);
    });

    test('query-basic', (t) async {
      await _setup();
      await _runset(_g('prepareQuery.basic'), _utility.prepareQuery);
    });

    test('body-basic', (t) async {
      await _setup();
      await _runset(_g('prepareBody.basic'), (dynamic ctx) {
        _fixctx(ctx);
        return _utility.prepareBody(ctx);
      });
    });

    test('findparam-basic', (t) async {
      await _setup();
      await _runset(_g('param.basic'), _utility.param);
    });

    test('fullurl-basic', (t) async {
      await _setup();
      await _runset(_g('makeUrl.basic'), _utility.makeUrl);
    });

    test('operator-basic', (t) async {
      await _setup();
      await _runset(_g('operator.basic'), (dynamic opmap) {
        return {
          'entity': opmap['entity'] ?? '_',
          'name': opmap['name'] ?? '_',
          'input': opmap['input'] ?? '_',
          'points': opmap['points'] ?? [],
        };
      });
    });

    test('options-basic', (t) async {
      await _setup();
      await _runset(_g('makeOptions.basic'), (dynamic vin) {
        final ctx = _utility.makeContext(
            {'options': vin['options'], 'config': vin['config']});
        ctx.client = _client;
        ctx.utility = _client.utility();
        return _utility.makeOptions(ctx);
      });
    });

    test('spec-basic', (t) async {
      await _setup();
      final sdkopts = vs.getpath(_spec, 'makeSpec.DEF.setup.a') ?? {};
      final specClient = ProjectNameSDK.test({}, sdkopts);
      await _runset(_g('makeSpec.basic'), (dynamic ctx) {
        ctx.client = specClient;
        ctx.options = specClient.options();
        return _utility.makeSpec(ctx);
      });
    });

    test('reqform-basic', (t) async {
      await _setup();
      await _runset(_g('transformRequest.basic'), _utility.transformRequest);
    });

    test('resform-basic', (t) async {
      await _setup();
      await _runset(_g('transformResponse.basic'), _utility.transformResponse);
    });

    test('resbasic-basic', (t) async {
      await _setup();
      await _runset(_g('resultBasic.basic'), (dynamic ctx) {
        _fixctx(ctx);
        return _utility.resultBasic(ctx);
      });
    });

    test('resheaders-basic', (t) async {
      await _setup();
      await _runset(_g('resultHeaders.basic'), (dynamic ctx) {
        // Header keys reach the pipeline lowercased by the transport.
        if (null != ctx.response && ctx.response.headers is Map) {
          final h = <String, dynamic>{};
          (ctx.response.headers as Map)
              .forEach((k, v) => h[k.toString().toLowerCase()] = v);
          ctx.response.headers = h;
        }
        return _utility.resultHeaders(ctx);
      });
    });

    test('resbody-basic', (t) async {
      await _setup();
      await _runset(_g('resultBody.basic'), (dynamic ctx) async {
        if (null != ctx.response && null == ctx.response.jsonFn) {
          final body = ctx.response.body;
          ctx.response.jsonFn = () => body;
        }
        return _utility.resultBody(ctx);
      });
    });

    test('request-basic', (t) async {
      await _setup();
      mockFetch(dynamic url, dynamic init) async => {
            'status': 200,
            'statusText': 'OK',
            'headers': {'content-type': 'application/json'},
            'json': () => {'id': 'res01'},
            'body': 'present',
          };
      final reqClient = ProjectNameSDK({
        'system': {'fetch': mockFetch}
      });
      final reqUtility = reqClient.utility();
      await _runset(_g('makeRequest.basic'), (dynamic ctx) async {
        ctx.client = reqClient;
        ctx.utility = reqUtility;
        ctx.options = reqClient.options();
        return reqUtility.makeRequest(ctx);
      });
    });

    test('response-basic', (t) async {
      await _setup();
      await _runset(_g('makeResponse.basic'), (dynamic ctx) async {
        _fixctx(ctx);
        if (null != ctx.response && null == ctx.response.jsonFn) {
          final body = ctx.response.body;
          ctx.response.jsonFn = () => body;
        }
        if (null != ctx.response && ctx.response.headers is Map) {
          final h = <String, dynamic>{};
          (ctx.response.headers as Map)
              .forEach((k, v) => h[k.toString().toLowerCase()] = v);
          ctx.response.headers = h;
        }
        return _utility.makeResponse(ctx);
      });
    });

    test('done-basic', (t) async {
      await _setup();
      await _runset(_g('done.basic'), (dynamic ctx) {
        _fixctx(ctx);
        return _utility.done(ctx);
      });
    });

    test('error-basic', (t) async {
      await _setup();
      await _runset(_g('makeError.basic'), (dynamic ctx, [dynamic err]) {
        _fixctx(ctx);
        return _utility.makeError(ctx, err);
      });
    });

    test('makePoint-single', (t) async {
      await _setup();
      final ctx = _makeCtx();
      final point = Point({
        'parts': ['items', '{id}'],
        'args': {'params': []},
        'params': [],
        'alias': {},
        'select': {},
        'active': true,
        'transform': {'req': null, 'res': null},
      });
      ctx.op.points = [point];

      final result = _utility.makePoint(ctx);
      ok(!iserr(result));
      equal(true, identical(ctx.point, point));
    });

    test('makeFetchDef', (t) async {
      await _setup();
      final ctx = _makeFullCtx();
      ctx.spec = _specOf({
        'base': 'http://localhost:8080',
        'prefix': '/api',
        'path': 'items/{id}',
        'suffix': '',
        'params': {'id': 'item01'},
        'query': {},
        'headers': {'content-type': 'application/json'},
        'method': 'GET',
        'step': 'start',
      });

      final fetchdef = _utility.makeFetchDef(ctx);
      ok(!iserr(fetchdef), 'should not be error');
      equal('GET', fetchdef['method']);
      ok(fetchdef['url'].toString().contains('/api/items/item01'));
      equal('application/json', fetchdef['headers']['content-type']);
      ok(null == fetchdef['body']);
    });

    test('makeFetchDef-with-body', (t) async {
      await _setup();
      final ctx = _makeFullCtx();
      ctx.spec = _specOf({
        'base': 'http://localhost:8080',
        'prefix': '',
        'path': 'items',
        'suffix': '',
        'params': {},
        'query': {},
        'headers': {},
        'method': 'POST',
        'step': 'start',
        'body': {'name': 'n0'},
      });

      final fetchdef = _utility.makeFetchDef(ctx);
      ok(!iserr(fetchdef));
      equal('POST', fetchdef['method']);
      ok(fetchdef['body'] is String);
      ok(fetchdef['body'].toString().contains('"n0"'));
    });

    test('fetcher-live', (t) async {
      await _setup();
      final calls = <Map<String, dynamic>>[];
      final liveClient = ProjectNameSDK({
        'system': {
          'fetch': (dynamic url, dynamic init) async {
            calls.add({'url': url, 'init': init});
            return {'status': 200, 'statusText': 'OK'};
          }
        }
      });
      final liveUtility = liveClient.utility();
      final ctx =
          liveUtility.makeContext({'opname': 'load'}, liveClient.rootctx);
      ctx.client = liveClient;

      final fetchdef = {'method': 'GET', 'headers': {}};
      final response =
          await liveUtility.fetcher(ctx, 'http://example.com/test', fetchdef);
      ok(!iserr(response));
      equal(1, calls.length);
      equal('http://example.com/test', calls[0]['url']);
    });

    test('fetcher-blocked-test-mode', (t) async {
      await _setup();
      final blockedClient = ProjectNameSDK({
        'system': {'fetch': (dynamic url, dynamic init) async => {}}
      });
      blockedClient.mode = 'test';

      final blockedUtility = blockedClient.utility();
      final ctx = blockedUtility
          .makeContext({'opname': 'load'}, blockedClient.rootctx);
      ctx.client = blockedClient;
      final fetchdef = {'method': 'GET', 'headers': {}};

      final result = await blockedUtility.fetcher(
          ctx, 'http://example.com/test', fetchdef);
      ok(iserr(result));
      ok(errmsg(result).contains('mode'));
    });

    test('makeError-no-throw', (t) async {
      await _setup();
      final ctx = _makeFullCtx();
      ctx.ctrl['throw'] = false;
      ctx.result = _resultOf({
        'ok': false,
        'resdata': {'id': 'safe01'},
      });

      final out =
          _utility.makeError(ctx, ctx.error('test_code', 'test message'));
      deepEqual(out, {'id': 'safe01'});
    });

    test('clean', (t) async {
      await _setup();
      final ctx = _makeFullCtx();
      final val = {'key': 'secret123', 'name': 'test'};
      final cleaned = _utility.clean(ctx, val);
      ok(null != cleaned);
    });
  });
}

// Helper functions for manual tests
dynamic _makeCtx([Map<String, dynamic>? overrides]) {
  final ctxmap = <String, dynamic>{'opname': 'load'};
  overrides?.forEach((k, v) => ctxmap[k] = v);
  return _utility.makeContext(ctxmap, _client.rootctx);
}

dynamic _makeFullCtx([Map<String, dynamic>? overrides]) {
  final ctx = _makeCtx(overrides);
  ctx.point = Point({
    'parts': ['items', '{id}'],
    'args': {
      'params': [
        {'name': 'id', 'reqd': true}
      ]
    },
    'params': ['id'],
    'alias': {},
    'select': {},
    'active': true,
    'relations': [],
    'transform': {'req': null, 'res': null},
  });
  ctx.match = {'id': 'item01'};
  ctx.reqmatch = {'id': 'item01'};
  return ctx;
}

dynamic _specOf(Map<String, dynamic> m) =>
    (_utility.makeContext({'spec': m}) as dynamic).spec;

dynamic _resultOf(Map<String, dynamic> m) =>
    (_utility.makeContext({'result': m}) as dynamic).result;
